### ISSUE TITLE
`orFail()` does not work if `err` is `undefined`

### ISSUE THREAD

**Issue by piranna** [Automattic/mongoose#7409]
**Do you want to request a *feature* or report a *bug*?**

Bug, in code or in docs.

**What is the current behavior?**

Related to https://github.com/Automattic/mongoose/issues/7280, according to the docs [`orFail()` throw an error if no documents match the given filter](https://mongoosejs.com/docs/api.html#query_Query-orFail). You can set what object to be thrown as an error or also set a function to generate it, but it doesn't specify what error object is thrown if none is specified. In that case, `orFail()` always resolve with an `undefined` value.

**If the current behavior is a bug, please provide the steps to reproduce.**

Reading the code it's thrown [whatever is on `err`](https://github.com/Automattic/mongoose/blob/4a4628718c2b9dbe87d7e136e4e21d59c6c9a600/lib/query.js#L3886), so it's throwing `undefined`, that's ok. Thing is that when callin `orFail()` it's being used `.map()` to add a new transform, and [when consuming `_transforms` in `_wrapThunkCallback()`](https://github.com/Automattic/mongoose/blob/4a4628718c2b9dbe87d7e136e4e21d59c6c9a600/lib/query.js#L3996), the thrown value is captured and set as error of the `cb()` callback... so being `undefined`, the callback thinks the operation was sucesful and resolved with the not provided `res` value, it's say, it resolve to `undefined`.

**What is the expected behavior?**

When `orFail()` don't have defined its `err` argument and no documents are found, it should reject to some value (maybe an empty `Error` object with no message or any other info?).

**Please mention your node.js, mongoose and MongoDB version.**

Node.js: 11.6.0
MongoDB: 3.6.4
Mongoose: 5.4.3

**Comment by piranna** [comment]
Or better than an empty Error object, add it some info about it's not found.

**Comment by piranna** [comment]
Maybe [DocumentNotFoundError](https://github.com/Automattic/mongoose/blob/ea3f10b8d5/lib/error/notFound.js) could be a good candidate?

**Comment by vkarpov15** [comment]
Good idea, we'll use DocumentNotFoundError. Thanks for finding this :+1:

Here's a script repro-ing this:

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7409`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const schema = new Schema({ name: String });
  const Model = mongoose.model('Test', schema);

  await Model.findOne().orFail(); // succeeds
}
```

**Comment by piranna** [comment]
> Thanks for finding this

You are welcome, glad you liked my idea too :-)

**Comment by piranna** [comment]
Thank you! :-D
